File: of06_从尾到头打印链表/Solution2.js
import { ListNode, ListUtils } from './utils_js/list.js';
import { ArrayUtils } from './utils_js/array.js';

/**
 * @param {ListNode} head
 * @return {number[]}
 */
var reversePrint = function(head) {
  let resArr = [];
  if (!head || head.length === 0) return resArr;

  let postOrder = (node) => {
    if (node.next) postOrder(node.next);
    resArr.push(node.val);
  }
  postOrder(head);
  return resArr;
};

function test(testName, head, expected) {
  let res = reversePrint(head);
  console.log('res = ', res)
  if (ArrayUtils.isEqualArray(res, expected))
    console.log(testName + ' success.');
  else
    console.log(testName + ' failed.');
}

let l1 = ListUtils.buildList([1,3,2]);
let expected1 = [2,3,1];
test('test1', l1, expected1);

let l2 = ListUtils.buildList([]);
let expected2 = [];
test('test2', l2, expected2);

File: of06_从尾到头打印链表/Solution2.go
package main

import (
	"fmt"
	"reflect"
	"slt/util_go/list"
)

type ListNode = list.ListNode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reversePrint(head *ListNode) []int {
	res := []int{}

	var dfs func(*ListNode)
	dfs = func(node *ListNode) {
		if node == nil {
			return
		}
		dfs(node.Next)
		res = append(res, node.Val)
	}
	dfs(head)

	return res
}

func test(testName string, head *ListNode, expected []int) {
	res := reversePrint(head)
	if reflect.DeepEqual(res, expected) {
		fmt.Println(testName + " success.")
	} else {
		fmt.Println(testName + " failed.")
	}
}

func main() {
	head1 := list.BuildList([]int{1,3,2})
	expected1 := []int{2,3,1}
	test("test1", head1, expected1)

}


// 输入一个链表的头节点，从尾到头反过来返回每个节点的值（用数组返回）。

// 示例 1：

// 输入：head = [1,3,2]
// 输出：[2,3,1]